leshan-core/src/main/java/org/eclipse/leshan/core/request/argument/Arguments.java

```diff
@@ -0,0 +1,314 @@
+     * @throws InvalidArgumentException if text has invalid format.
+     */
+    static Arguments parse(String arglist) throws InvalidArgumentException {
+        ArgumentsBuilder builder = builder();
+
+        if (arglist != null && !arglist.isEmpty()) {
+            int beginProbe = 0;
+            int validBegin = 0;
+            while (true) {
+                int separatorIndex = arglist.indexOf(',', beginProbe);
+                if (separatorIndex == -1) {
+                    break;
+                }
+                String argumentText = arglist.substring(validBegin, separatorIndex);
+
+                ArgumentParser argumentParser = new ArgumentParser(argumentText, arglist);
+                if (argumentParser.isValid()) {
+                    Argument argument = argumentParser.parse();
+                    builder.addArgument(argument);
+                    validBegin = separatorIndex + 1;
+                }
+                beginProbe = separatorIndex + 1;
+            }
+
+            String argumentText = arglist.substring(beginProbe);
+            ArgumentParser argumentParser = new ArgumentParser(argumentText, arglist);
+            Argument argument = argumentParser.parse();
+            builder.addArgument(argument);
+        }
+
+        // Use builder capability to validate arguments uniqueness.
+        return builder.build();
+    }
+
+    private static class ArgumentParser {
+        private final String digitPart;
+        private final String valueDecorated;
+        private final String content;
+
+        public ArgumentParser(String argument, String content) {
+            this.content = content;
+            String[] keyValue = argument.split("=", 2);
+            digitPart = keyValue[0];
+            valueDecorated = keyValue.length == 1 ? null : keyValue[1];
+        }
+
+        private Argument parse() throws InvalidArgumentException {
+            if (!isValid()) {
+                throw new InvalidArgumentException("Unable to parse Arguments [%s]", content);
+            }
+            int digit;
+            try {
+                digit = Integer.parseInt(digitPart);
+            } catch (NumberFormatException e) {
+                throw new InvalidArgumentException(e, "Unable to parse Arguments [%s] with digit [%s]", content, digitPart);
+            }
+            String value = null;
+            if (valueDecorated != null) {
+                value = StringUtils.removeEnd(StringUtils.removeStart(valueDecorated, "'"), "'");
+            }
+
+            return new Argument(digit, value);
+        }
+
+        private boolean isValid() {
+            if (digitPart.length() != 1) {
+                return false;
+            }
+
+            if (valueDecorated != null) {
+                return valueDecorated.length() >= 2 && valueDecorated.charAt(0) == '\''
+                        && valueDecorated.charAt(valueDecorated.length() - 1) == '\'';
+            }
+            return true;
+        }
```

Comment:
About isValid, I understood that you plan to inline the checks in `parse` (https://github.com/eclipse/leshan/pull/1148#discussion_r758264083)
Benefits will be you will be able to create more detailed error mesage

Reply:
Method ArgumentParser.isValid() has two functions:
* standard checking while parsing (allow throwing an error while parsing)
* pre-validation part of text to detect if it's able to parse into argument (it helps to deal with commas inside quotes)

So I can't inline this method because it's used in both places. For more detailed error messages I could use return codes to differentiate between validation errors. In first usage I could customise error messages, second usage can ignore return code because it doesn't matter.